feat: local client

ryanolson opened the pull request:
## Summary

Implements LocalClient for direct, in-process engine access without network overhead. Enables type-safe engine invocation for testing and local development scenarios where network latency is not desired.

## Key Technical Changes

### 1. StreamData Trait for Non-Sync Response Types
- **Problem**: `ManyOut<T>` streams don't satisfy `Sync` requirement for local registry storage
- **Solution**: New `StreamData` trait requiring only `Send + 'static` (not `Sync`)
- **Implementation**: `StreamAnyEngineWrapper` provides type erasure for heterogeneous storage
- **Benefit**: Enables streaming responses in local client scenarios

```rust
// New trait for stream-compatible responses
pub trait StreamData: Send + 'static {}
impl<T: Send + 'static> StreamData for T {}

// Type-erased wrapper for storage
pub struct StreamAnyEngineWrapper<Req, Resp, E> {
    engine: Arc<dyn AsyncEngine<Req, Resp, E>>,
}
```

### 2. LocalClient Implementation
- **Type-safe downcasting**: Generic client with proper error handling for mismatched types
- **Direct invocation**: Bypasses network stack (etcd/NATS) for zero-latency calls
- **Automatic registration**: Engines registered during endpoint lifecycle management
- **Registry lookup**: Efficient HashMap-based engine resolution using entity descriptors

```rust
// Usage example
let local_client: LocalClient<SingleIn<i32>, ManyOut<Annotated<i32>>, anyhow::Error> = 
    endpoint.local_client().await?;
let response = local_client.generate(request).await?;
```

### 3. Endpoint Lifecycle Management
- **Resource leak fix**: Moved local engine registration from `create()` to `start()` to prevent leaks when EndpointInstance is dropped without starting
- **Background execution**: Added `start_background()` methods using `CriticalTaskExecutionHandle` for proper cleanup and failure propagation
- **Graceful shutdown**: Background endpoints register with graceful shutdown tracker and unregister on completion

### 4. Internal Namespace Handling
- **Validation fix**: Added conditional logic to use `NamespaceDescriptor::new_internal()` for namespaces starting with "_"
- **Entity descriptors**: Runtime v2 descriptors provide fluent builders with proper validation for namespaces, components, and endpoints

### 5. Health Check Integration
- **Endpoint-specific notifications**: Added health check timer reset notifications when streams complete
- **SystemHealth integration**: Proper registration and cleanup of health check targets

## Architecture

```
┌─────────────────┐    ┌──────────────────┐    ┌─────────────────┐
│   LocalClient   │────│  Local Registry  │────│  Actual Engine  │
│  (Type-safe)    │    │   (HashMap)      │    │  (In-process)   │
└─────────────────┘    └──────────────────┘    └─────────────────┘
         │                        │                       │
         │                        │                       │
         ▼                        ▼                       ▼
   Direct method call      Fast lookup           Zero network overhead
```

## Files Changed

### Core Implementation
- `lib/runtime/src/engine.rs` - StreamData trait and StreamAnyEngineWrapper
- `lib/runtime/src/component/local_client.rs` - LocalClient with type-safe downcasting
- `lib/runtime/src/component/endpoint.rs` - Lifecycle management and resource leak fixes
- `lib/runtime/src/distributed.rs` - Local registry and background endpoint tracking

### Integration Points  
- `lib/runtime/src/pipeline/network/ingress/push_handler.rs` - Engine registration and health check notifications
- `lib/runtime/src/pipeline/network.rs` - PushWorkHandler trait extensions
- `lib/runtime/src/v2/entity.rs` - Runtime v2 entity descriptors with validation

### Examples and Tests
- `examples/local_client_demo.rs` - Comprehensive demo with proper Annotated struct usage
- `lib/runtime/src/component/local_client/tests.rs` - Type-safety and error handling tests

## Testing Strategy

### Automated Tests
- **Type safety**: Verified LocalClient rejects mismatched engine types
- **Stream handling**: Tested `ManyOut<Annotated<T>>` with proper field usage (data, id, event, comment)
- **Error propagation**: Validated proper error handling throughout the stack
- **Registry integration**: Confirmed engines are properly registered and accessible

### Manual Verification
- **Clippy compliance**: All warnings resolved (unused variables, complex types)
- **Resource cleanup**: Verified no leaks when EndpointInstance is dropped
- **Background execution**: Confirmed proper cleanup with CriticalTaskExecutionHandle

## Migration Notes

### For Existing Code
- Existing endpoint usage remains unchanged
- LocalClient is opt-in via `endpoint.local_client().await?` 
- Per-service toggle available to disable local registry if needed

### Performance Considerations
- **Zero network latency**: Direct in-process calls
- **Minimal overhead**: HashMap lookup + type casting
- **Memory efficient**: Shared Arc references, no data copying

## Addresses Review Feedback

1. **Resource leaks**: Fixed by moving registration from create() to start()
2. **Internal namespace handling**: Added proper validation for "_" prefixed namespaces  
3. **Documentation coverage**: Comprehensive docs added (>80% coverage target)
4. **Background execution**: Implemented using CriticalTaskExecutionHandle pattern
5. **Health check integration**: Proper endpoint-specific timer reset notifications

## Risk Assessment

- **Low risk**: Additive feature, doesn't modify existing network paths
- **Type safety**: Compile-time guarantees prevent runtime type errors  
- **Graceful degradation**: Falls back to network calls if local engine unavailable
- **Resource management**: RAII patterns ensure proper cleanup

ryanolson committed:
fix: use is_some_and instead of map_or in v2::entity descriptor

- Remove unnecessary clippy allow annotation
- Use is_some_and which is available in Rust 1.70+ (we're on 1.89)
- Clean up code to follow clippy recommendations

ryanolson committed:
feat: implement local client registry for direct engine access

- Add v2::entity module for type-safe entity descriptors
- Implement LocalClient for bypassing network overhead
- Add local engine registry to DistributedRuntime
- Auto-register engines during endpoint creation
- Add comprehensive tests and examples

This implementation allows direct local engine invocation without the
overhead of etcd watching, instance discovery, or network layers. Engines
are automatically registered when endpoints are created and can be
accessed via LocalClient using the v2::entity descriptor system for
consistent addressing.

ryanolson committed:
refactor: improve LocalClient API and fix namespace hierarchy

- Fix namespace hierarchy extraction to walk parent chain properly
- Make registry methods internal (pub(crate)) for better encapsulation
- Add ServiceBuilder support with with_local_registry/without_local_registry
- Check enable_local_registry flag during endpoint creation (enabled by default)
- Add endpoint.local_client() convenience method for consistency
- Update tests to use new API without direct registry access
- Fix duplicate namespace hierarchy helper functions

This improves the LocalClient API to be more consistent with existing
patterns while properly handling hierarchical namespaces and providing
better encapsulation of internal implementation details.

ryanolson committed:
refactor: separate endpoint creation from starting and improve LocalClient encapsulation

- Split EndpointConfigBuilder into create() and start() phases
- create() handles setup including local engine registration
- start() handles network serving
- Maintain backwards compatibility with existing start() -> create().start()
- Remove public register_local_engine function, now internal
- Fix namespace hierarchy extraction to properly walk parent chains
- Update tests to use new create/start pattern (tests need further API alignment)

The core implementation compiles successfully, though tests need additional work
to align with current pipeline API requirements.

Signed-off-by: Ryan Olson <rolson@nvidia.com>

ryanolson committed:
feat: implement StreamData trait for local client engine registry

This change addresses trait bound conflicts in the local client system by introducing
a new StreamData trait that only requires Send (not Sync) for response types.

Key changes:
- Add StreamData trait for response types that don't need Sync
- Create StreamAnyEngineWrapper for type erasure with StreamData responses
- Update LocalClient to use StreamData constraint for response types
- Add downcast_stream() method for StreamData-compatible downcasting
- Fix clippy warnings about complex types and unused variables
- Rewrite local_client_demo.rs to use new endpoint builder pattern

This allows ManyOut<T> types (Pin<Box<dyn AsyncEngineStream<T>>>) to work
properly with the local engine registry, since streams are consumed
sequentially and don't need to be shared between threads.

Signed-off-by: Ryan Olson <rolson@nvidia.com>

ryanolson committed:
merge: integrate main branch with local client features

Merge main branch to bring in health check functionality while
preserving local client registry features.

Key changes resolved:
- Combined local registry fields (engine, any_engine) with health check
  field (endpoint_health_check_notifier) in Ingress struct
- Merged PushWorkHandler trait methods: both as_any_engine() for local
  registry and set_endpoint_health_check_notifier() for health checks
- Integrated health check registration logic into endpoint create/start
  pattern while maintaining the refactored separation
- Added health_check_payload field to EndpointInstance and EndpointConfig

Both feature sets now work together:
- Local client can access engines directly without network overhead
- Health check system monitors endpoint availability and resets timers
- Endpoints support both local registry and health monitoring

All library unit tests pass. Integration tests require external services.

Signed-off-by: Ryan Olson <rolson@nvidia.com>

grahamking commented:
Is this what `dynamo-run in=http out=mistralrs|llamacpp` does, which it calls `StaticFull` and `StaticCore`? Or does this replace NATS in a distributed setup?

ryanolson committed:
Merge branch 'main' into ryan/dyn-986-local-client

ryanolson committed:
fix: address resource leak and internal namespace handling

- Move local engine registration from create() to start() to prevent leaks
- Add conditional internal namespace handling for names starting with '_'
- Implement start_background() with CriticalTaskExecutionHandle
- Add background endpoint registry for lifecycle management
- Ensure proper cleanup when endpoints complete or fail

Key changes:
* EndpointInstance.local_engine_info now stores prepared registration data
* Registration happens in start() method, avoiding leaks if dropped
* Internal namespaces use NamespaceDescriptor::new_internal()
* Background endpoints tracked in DistributedRuntime registry
* CriticalTaskExecutionHandle manages endpoint lifecycle automatically

Signed-off-by: Ryan Olson <rolson@nvidia.com>
